ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp

```diff
@@ -24,231 +23,230 @@
-			case BATTERY_TYPE_UNKNOWN:
-				return gp_unknown;
-				break;
-			default:
-				return -2;
-		}
-    } else {
-        return -1;
+  XINPUT_BATTERY_INFORMATION batteryInformation = {};
+
+  DWORD dwResult = XInputGetBatteryInformation(
+      device, BATTERY_DEVTYPE_GAMEPAD, &batteryInformation);
+
+  if (dwResult == ERROR_SUCCESS) {
+    switch (batteryInformation.BatteryType) {
+      case BATTERY_TYPE_DISCONNECTED:
+        return gp_disconnected;
+      case BATTERY_TYPE_WIRED:
+        return gp_wired;
+      case BATTERY_TYPE_ALKALINE:
+        return gp_alkaline;
+      case BATTERY_TYPE_NIMH:
+        return gp_nimh;
+      case BATTERY_TYPE_UNKNOWN:
+      default:
+        return gp_unknown;
     }
+  }
+
+  return 0;
 }
-	
+
 int gamepad_get_battery_charge(int device) {
-	XINPUT_BATTERY_INFORMATION batteryInformation;
-    // Zeroise the state
-    ZeroMemory(&batteryInformation, sizeof(XINPUT_BATTERY_INFORMATION));
-    // Get the state
-    DWORD Result = XInputGetBatteryInformation(device, XINPUT_DEVTYPE_GAMEPAD, &batteryInformation);
-	if (Result == ERROR_SUCCESS) {
-        switch (batteryInformation.BatteryLevel) {
-			case BATTERY_LEVEL_EMPTY:
-				return gp_empty;
-				break;
-			case BATTERY_LEVEL_LOW:
-				return gp_low;
-				break;
-			case BATTERY_LEVEL_MEDIUM:
-				return gp_medium;
-				break;
-			case BATTERY_LEVEL_FULL:
-				return gp_full;
-				break;
-			default:
-				return -2;
-		}
-    } else {
-        return -1;
+  XINPUT_BATTERY_INFORMATION batteryInformation = {};
+
+  DWORD dwResult = XInputGetBatteryInformation(
+      device, XINPUT_DEVTYPE_GAMEPAD, &batteryInformation);
+
+  if (dwResult == ERROR_SUCCESS) {
+    switch (batteryInformation.BatteryLevel) {
+      case BATTERY_LEVEL_EMPTY:
+        return gp_empty;
+      case BATTERY_LEVEL_LOW:
+        return gp_low;
+      case BATTERY_LEVEL_MEDIUM:
+        return gp_medium;
+      case BATTERY_LEVEL_FULL:
+        return gp_full;
+      default:
+        break;
     }
+  }
+
+  return 0;
 }
 
 float gamepad_get_button_threshold(int device) {
-	
+
 }
-	
+
 void gamepad_set_button_threshold(int device, float threshold) {
 
 }
-	
-void gamepad_set_vibration(int device, float left, float right) {
-	XINPUT_STATE controllerState;
-    // Create a Vibraton State
-    XINPUT_VIBRATION vibration;
 
-    // Zeroise the Vibration
-    ZeroMemory(&vibration, sizeof(XINPUT_VIBRATION));
+void gamepad_set_axis_deadzone(int device, float deadzone) {
+
+}
+
+void gamepad_set_vibration(int device, float left, float right) {
+  XINPUT_VIBRATION vibration = {};
 
-    // Set the Vibration Values
-    vibration.wLeftMotorSpeed = left * 65535;
-    vibration.wRightMotorSpeed = right * 65535;
+  vibration.wLeftMotorSpeed = left * 65535;
+  vibration.wRightMotorSpeed = right * 65535;
 
-    // Vibrate the controller
-    XInputSetState(device, &vibration);
+  XInputSetState(device, &vibration);
 }
-	
+
 int gamepad_axis_count(int device) {
-	return 2;
+  return 4;
```

Comment:
When it comes to XInput Josh, 4 controllers is the limit.
http://gaming.stackexchange.com/a/196367

For DirectInput you can have as many as you want.